### PR TITLE
Slight updates and changes to unittest emulations. Special case works…

### PR DESCRIPTION
**Pull request by redheadedrod**

### CODE REVIEW

File: envi/archs/arm/disasm.py [hunk 1]
```diff
@@ -209,7 +209,11 @@ def p_dp_imm_shift(opval, va):
 
     opcode = (IENC_DP_IMM_SHIFT << 16) + ocode
     if sflag > 0:
-        iflags = IF_PSR_S
+        # IF_PSR_S_SIL is silent s for tst, teq, cmp cmn
+        if ocode in [8, 9, 10, 11]:
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: ewwww... magic numbers!  please define and use names here.

File: envi/archs/arm/disasm.py [hunk 2]
```diff
@@ -548,12 +552,17 @@ def p_dp_reg_shift(opval, va):
 
     opcode = (IENC_DP_REG_SHIFT << 16) + ocode
     if sflag > 0:
-        iflags = IF_PSR_S
+        # IF_PSR_S_SIL is silent s for tst, teq, cmp cmn
+        if ocode in [8, 9, 10, 11]:
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: see above.  same here.

File: envi/archs/arm/disasm.py [hunk 3]
```diff
@@ -610,7 +619,11 @@ def p_dp_imm(opval, va):
 
     opcode = (IENC_DP_IMM << 16) + ocode
     if sflag > 0:
-        iflags = IF_PSR_S
+        # IF_PSR_S_SIL is silent s for tst, teq, cmp cmn
+        if ocode in [8, 9, 10, 11]:
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: ditto.

File: envi/archs/arm/disasm.py [hunk 4]
```diff
@@ -1455,7 +1468,7 @@ def render(self, mcanv):
         daib_flags = self.iflags & IF_DAIB_MASK
         if self.iflags & IF_L:
             mnem += 'l'
-        elif self.iflags & IF_PSR_S:
+        elif self.iflags & IF_PSR_S and not (self.iflags & IF_PSR_S_SIL):
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: this works.  i prefer the following approach (for speed):

defined in global module context:
S_FLAG_MASK = IF_PSR_S | IF_PSR_S_SIL

in the function:
    elif (self.iflag & S_FLAG_MASK) == IF_PSR_S:

the performance of the math is faster than two distinct checks.

File: envi/archs/arm/disasm.py [hunk 5]
```diff
@@ -1523,7 +1536,7 @@ def __repr__(self):
         daib_flags = self.iflags & IF_DAIB_MASK
         if self.iflags & IF_L:
             mnem += 'l'
-        elif self.iflags & IF_PSR_S:
+        elif (self.iflags & IF_PSR_S) and not (self.iflags & IF_PSR_S_SIL):
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: ditto

File: envi/archs/arm/disasm.py [hunk 6]
```diff
@@ -2623,14 +2636,14 @@ def doDecode(self, va, opval, bytez, offset):
                     if (opval & mask) == val:
                         enc = penc
                         break
-
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: i liked this newline :)

File: envi/archs/arm/emu.py [hunk 7]
```diff
@@ -777,8 +777,14 @@ def i_adc(self, op):
         
         dsize = op.opers[0].tsize
         ssize = op.opers[1].tsize
-        #FIXME needs sflag
         Carry = self.getFlag(PSR_C_bit)
+        #FIXME is this right?
+        if op.iflags & IF_PSR_S > 1:
```
Comment (atlas0fd00m) [inline envi/archs/arm/emu.py]: Sflag = bool(op.iflags & IF_PSR_S)

File: envi/tests/test_arch_arm.py [hunk 8]
```diff
@@ -910,11 +888,10 @@ def validateEmulation(self, emu, op, setters, tests):
         ## defaults
         emu.setRegister(REG_R3, 0x414141)
         emu.setRegister(REG_R4, 0x444444)
-        emu.setRegister(REG_R5, 0x454545)
+        emu.setRegister(REG_R5, 0x10)
```
Comment (atlas0fd00m) [inline envi/tests/test_arch_arm.py]: ?

File: envi/archs/arm/disasm.py [hunk 9]
```diff
@@ -517,7 +532,7 @@ def p_load_store_word_ubyte(opval, va):
         iflags = IF_B
 
     if (pubwl & 0x12) == 2:
-        iflags |= IF_T
+        iflags |= IF_T
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: whspc

File: envi/archs/arm/disasm.py [hunk 10]
```diff
@@ -1548,9 +1565,9 @@ def __repr__(self):
                 mnem += 'd'
             if self.iflags & IF_B:
                 mnem += 'b'
-            if self.iflags & IF_H:
+            if self.iflags & IF_H: 
                 mnem += 'h'
-            elif self.iflags & IF_T:
+            if self.iflags & IF_T: #removed el
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: is this the same IF_T meaning as when paired with IF_H?

File: envi/archs/arm/disasm.py [hunk 11]
```diff
@@ -1548,9 +1565,9 @@ def __repr__(self):
                 mnem += 'd'
             if self.iflags & IF_B:
                 mnem += 'b'
-            if self.iflags & IF_H:
+            if self.iflags & IF_H:
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: whspc

File: envi/archs/arm/disasm.py (hunk 10 above)
Comment (redheadedrod) [inline envi/archs/arm/disasm.py]: From what I read unless mistaken STR is same as STRT but the 'T' variant will run with user mode permissions. Same for STRH and STRHT. The ONLY issue I saw was that there are a couple literal versions of LDR statements and not all of them supported the T variant as commented in the code before this code. I looked through the reference to find the T variants and how they compared. I can double check if necessary but that is what my impression was. And this version of the code matches all of the tests.

File: envi/archs/arm/disasm.py (hunk 9 above)
Comment (redheadedrod) [inline envi/archs/arm/disasm.py]: Should have been fixed in last commit. Caught this after I committed already.

File: envi/archs/arm/disasm.py (hunk 11 above)
Comment (redheadedrod) [inline envi/archs/arm/disasm.py]: ditto on this one

File: envi/archs/arm/__init__.py [hunk 12]
```diff
@@ -56,6 +56,7 @@ class ThumbModule(envi.ArchitectureModule):
     archVersion = ('ARMv7A')
 
     def __init__(self, name='ARMv7A'):
+        self.archVersion = name
```
Comment (atlas0fd00m) [inline envi/archs/arm/__init__.py]: i'm not sure this will do what you think it will do :)  but we'll let it go for now.

**Pull request merged**